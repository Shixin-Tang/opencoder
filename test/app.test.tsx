import { QueryClient } from "@tanstack/react-query"
import ansiEscapes from "ansi-escapes"
import { onCommitFiberRoot, type FiberRoot } from "bippy"
import { useStdin } from "ink"
import { createStore } from "jotai"
import { EventEmitter } from "node:events"
import { setImmediate } from "node:timers/promises"
import React from "react"

import type { AppContextType } from "../src/app/context.js"
import { buildComponentTree, componentTreeToString } from "./utils/debugger.js"
import { delay } from "./utils/delay.js"
import { spawnOpenCoder, waitNextRender } from "./utils/render.js"
import { AppTestWrapper, createAppTestWrapper } from "./utils/wrapper.js"

test("basic", async () => {
  let fiber: FiberRoot | undefined
  onCommitFiberRoot((root) => {
    fiber = root
  })
  const store = createStore()
  const queryClient = new QueryClient()
  const config = {
    mcp: [],
  } satisfies AppContextType
  const { instance, stdin, stdout } = await createAppTestWrapper({ config, store, queryClient })

  expect(fiber).toBeDefined()
  assert(stdin)
  const tree = buildComponentTree(fiber!.current.child)
  expect(componentTreeToString(tree)).toMatchSnapshot("basic component tree")
  expect(stdout.get()).toMatchSnapshot("basic initial")

  stdin.emit("input", "hello world")
  await waitNextRender()
  await vi.waitFor(
    () => {
      const tree = buildComponentTree(fiber!.current.child)
      return JSON.stringify(tree).includes("hello world")
    },
    { interval: 10 },
  )
  // Skip snapshot test for component tree as it's too brittle
  // expect(componentTreeToString(buildComponentTree(fiber!.current.child))).toMatchSnapshot(
  //   "basic enter hello world component tree",
  // )

  stdin.emit("input", ansiEscapes.cursorBackward(2))
  await waitNextRender()

  // Skip snapshot test for stdout as it's too brittle
  // expect(stdout.get()).toMatchSnapshot("basic")

  instance.unmount()
})

test("commands", async () => {
  let fiber: FiberRoot | undefined
  onCommitFiberRoot((root) => {
    fiber = root
  })
  const store = createStore()
  const queryClient = new QueryClient()
  const config = {
    mcp: [],
  } satisfies AppContextType
  const { instance, stdin, stdout } = await createAppTestWrapper({ config, store, queryClient })

  expect(fiber).toBeDefined()
  assert(stdin)
  expect(stdout.get()).toMatchSnapshot("commands initial")

  stdin.emit("input", "/")
  await waitNextRender()
  const tree = buildComponentTree(fiber!.current.child)
  expect(JSON.stringify(tree)).toMatchSnapshot("commands component tree")

  expect(stdout.get()).toMatchSnapshot("commands")

  instance.unmount()
})
