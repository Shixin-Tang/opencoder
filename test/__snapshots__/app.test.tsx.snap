// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > basic 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;

exports[`basic > basic component tree 1`] = `
"<App2>
  <AppTestWrapper>
    <Provider state={}>
      <Unknown>
        <QueryClientProvider state={}>
          <Unknown>
            <AppProvider>
              <Unknown>
                <App state={}>
                  <Box flexDirection="column" flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                    <ink-box>
                      <Box flexDirection="column" gap={2} display="flex" flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                        <ink-box>
                          <Chat state={}>
                            <Box flexDirection="column" gap={0} flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                              <ink-box>
                                <Unknown input="" isDisabled={false} isLoading={false} state={}>
                                  <Box flexDirection="column" flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                                    <ink-box>
                                      <Box alignItems="flex-start" justifyContent="flex-start" borderColor="#888" borderDimColor={true} borderStyle="round" marginTop={1} width="100%" flexDirection="row" flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                                        <ink-box>
                                          <Box alignItems="center" alignSelf="flex-start" flexWrap="nowrap" justifyContent="center" width={3} flexDirection="row" flexGrow={0} flexShrink={1}>
                                            <ink-box>
                                              <Text children=">">
                                                <ink-text children=">">
                                                  <Unknown _0=">">
                                                </ink-text>
                                              </Text>
                                            </ink-box>
                                          </Box>
                                          <Box paddingRight={1} flexWrap="nowrap" flexDirection="row" flexGrow={0} flexShrink={1}>
                                            <ink-box>
                                              <TextInput multiline={true} value="" placeholder="" columns={74} isDimmed={false} disableCursorMovementForUpDownKeys={false} cursorOffset={0} state={}>
                                                <Text wrap="truncate-end" dimColor={false} children=" ">
                                                  <ink-text children=" ">
                                                    <Unknown _0=" ">
                                                  </ink-text>
                                                </Text>
                                              </TextInput>
                                            </ink-box>
                                          </Box>
                                        </ink-box>
                                      </Box>
                                      <Box flexDirection="row" justifyContent="space-between" paddingX={2} paddingY={0} flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                                        <ink-box>
                                          <Box justifyContent="flex-start" gap={1} flexDirection="row" flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                                            <ink-box>
                                              <Unknown $$typeof=undefined>
                                                <Text dimColor={true} children="/ for commands">
                                                  <ink-text children="/ for commands">
                                                    <Unknown _0="/" _1=" " _2="f" _3="o" _4="r" _5=" " _6="c" _7="o" _8="m" _9="m" _10="a" _11="n" _12="d" _13="s">
                                                  </ink-text>
                                                </Text>
                                              </Unknown>
                                            </ink-box>
                                          </Box>
                                          <Box justifyContent="flex-end" gap={1} flexDirection="row" flexWrap="nowrap" flexGrow={0} flexShrink={1}>
                                            <ink-box>
                                              <Text dimColor={true} children="alt + ⏎ for newline">
                                                <ink-text children="alt + ⏎ for newline">
                                                  <Unknown _0="a" _1="l" _2="t" _3=" " _4="+" _5=" " _6="⏎" _7=" " _8="f" _9="o" _10="r" _11=" " _12="n" _13="e" _14="w" _15="l" _16="i" _17="n" _18="e">
                                                </ink-text>
                                              </Text>
                                              <Text dimColor={true}>
                                                <ink-text>
                                                  <Unknown _0="·">
                                                  <Unknown _0="0" _1=" " _2="t" _3="o" _4="k" _5="e" _6="n" _7="s">
                                                </ink-text>
                                              </Text>
                                            </ink-box>
                                          </Box>
                                        </ink-box>
                                      </Box>
                                      <Unknown state={"retryLane":0}>
                                        <Unknown state={"baseLanes":32}>
                                        <Unknown>
                                      </Unknown>
                                    </ink-box>
                                  </Box>
                                </Unknown>
                              </ink-box>
                            </Box>
                          </Chat>
                        </ink-box>
                      </Box>
                      <ToolConfirmationDialog state={}>
                    </ink-box>
                  </Box>
                </App>
              </Unknown>
            </AppProvider>
          </Unknown>
        </QueryClientProvider>
      </Unknown>
    </Provider>
  </AppTestWrapper>
</App2>"
`;

exports[`basic > basic initial 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;

exports[`commands > commands 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ > /                                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  /sync    Sync codebase to the codebase index, stored in .coder/embeddings"
`;

exports[`commands > commands component tree 1`] = `"[{"name":"App2","props":{},"state":{},"children":[{"name":"AppTestWrapper","props":{},"state":{},"children":[{"name":"Provider","props":{},"state":{},"children":[{"name":"Unknown","props":{},"state":{},"children":[{"name":"QueryClientProvider","props":{},"state":{},"children":[{"name":"Unknown","props":{},"state":{},"children":[{"name":"AppProvider","props":{},"state":{},"children":[{"name":"Unknown","props":{},"state":{},"children":[{"name":"App","props":{},"state":{},"children":[{"name":"Box","props":{"flexDirection":"column","flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Box","props":{"flexDirection":"column","gap":2,"display":"flex","flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Chat","props":{},"state":{},"children":[{"name":"Box","props":{"flexDirection":"column","gap":0,"flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Unknown","props":{"input":"/","isDisabled":false,"isLoading":false},"state":{},"children":[{"name":"Box","props":{"flexDirection":"column","flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Box","props":{"alignItems":"flex-start","justifyContent":"flex-start","borderColor":"#888","borderDimColor":true,"borderStyle":"round","marginTop":1,"width":"100%","flexDirection":"row","flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Box","props":{"alignItems":"center","alignSelf":"flex-start","flexWrap":"nowrap","justifyContent":"center","width":3,"flexDirection":"row","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Text","props":{"children":">"},"state":{},"children":[{"name":"ink-text","props":{"children":">"},"state":{},"children":[{"name":"Unknown","props":{"0":">"},"state":{},"children":[]}]}]}]}]},{"name":"Box","props":{"paddingRight":1,"flexWrap":"nowrap","flexDirection":"row","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"TextInput","props":{"multiline":true,"value":"/","placeholder":"","columns":74,"isDimmed":false,"disableCursorMovementForUpDownKeys":true,"cursorOffset":1},"state":{},"children":[{"name":"Text","props":{"wrap":"truncate-end","dimColor":false,"children":"/ "},"state":{},"children":[{"name":"ink-text","props":{"children":"/ "},"state":{},"children":[{"name":"Unknown","props":{"0":"/","1":" "},"state":{},"children":[]}]}]}]}]}]}]}]},{"name":"Box","props":{"flexDirection":"row","justifyContent":"space-between","paddingX":2,"paddingY":0,"flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Box","props":{"flexDirection":"column","flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Box","props":{"flexDirection":"row","flexWrap":"nowrap","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Box","props":{"width":9,"flexWrap":"nowrap","flexDirection":"row","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Text","props":{"color":"#b1b9f9","dimColor":false},"state":{},"children":[{"name":"ink-text","props":{},"state":{},"children":[{"name":"Unknown","props":{"0":"/"},"state":{},"children":[]},{"name":"Unknown","props":{"0":"s","1":"y","2":"n","3":"c"},"state":{},"children":[]}]}]}]}]},{"name":"Box","props":{"width":67,"paddingLeft":0,"flexWrap":"nowrap","flexDirection":"row","flexGrow":0,"flexShrink":1},"state":{},"children":[{"name":"ink-box","props":{},"state":{},"children":[{"name":"Text","props":{"color":"#b1b9f9","dimColor":false,"wrap":"wrap"},"state":{},"children":[{"name":"ink-text","props":{},"state":{},"children":[{"name":"Text","props":{"dimColor":false},"state":{},"children":[{"name":"ink-text","props":{},"state":{},"children":[{"name":"Unknown","props":{"0":"S","1":"y","2":"n","3":"c","4":" ","5":"c","6":"o","7":"d","8":"e","9":"b","10":"a","11":"s","12":"e","13":" ","14":"t","15":"o","16":" ","17":"t","18":"h","19":"e","20":" ","21":"c","22":"o","23":"d","24":"e","25":"b","26":"a","27":"s","28":"e","29":" ","30":"i","31":"n","32":"d","33":"e","34":"x","35":",","36":" ","37":"s","38":"t","39":"o","40":"r","41":"e","42":"d","43":" ","44":"i","45":"n","46":" ","47":".","48":"c","49":"o","50":"d","51":"e","52":"r","53":"/","54":"e","55":"m","56":"b","57":"e","58":"d","59":"d","60":"i","61":"n","62":"g","63":"s"},"state":{},"children":[]}]}]}]}]}]}]}]}]}]}]}]}]},{"name":"Unknown","props":{},"state":{"retryLane":0},"children":[{"name":"Unknown","props":{},"state":{"baseLanes":32},"children":[]},{"name":"Unknown","props":{},"state":{},"children":[]}]}]}]}]}]}]}]}]}]},{"name":"ToolConfirmationDialog","props":{},"state":{},"children":[]}]}]}]}]}]}]}]}]}]}]}]}]"`;

exports[`commands > commands initial 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;
